[Impeller] `vk::CommandPool` resets via `ResourceManager`

matanlurey opened the pull request:
Fixes https://github.com/flutter/flutter/issues/133198.

Major changes:

- `CommandPoolVK` is now created indirectly through `CommandPoolRecyclerVK`.
- Via destructor, `CommandPoolVK` resets on a background thread via `ResourceManager`.
- Removed all of the code trying to reuse/reset individual command buffers (no longer needed).
- After every frame, the current/active command pool is recycled.

---

_Tests secured._

jonahwilliams commented on impeller/renderer/backend/vulkan/command_pool_vk.cc:
```diff
@@ -3,166 +3,190 @@
-static std::unordered_map<const ContextVK*,
-                          std::vector<std::weak_ptr<CommandPoolVK>>>
-    g_all_pools IPLR_GUARDED_BY(g_all_pools_mutex);
+  // These are retained otherwise it is a Vulkan thread-safety violation
+  // because the buffers and the originating pool do not belong to the same
+  // thread.
```
nit: these are retained because the destructor of the C++ UniqueCommandBuffer wrapper type will attempt to reset the cmd buffer, and doing so may be a thread safety violation as this may happen on the fence waiter thread.

matanlurey replied:
Thanks. Changed!

chinmaygarde commented on impeller/renderer/backend/vulkan/test/mock_vulkan.cc:
```diff
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 #include "impeller/renderer/backend/vulkan/test/mock_vulkan.h"
+#include "vulkan/vulkan_core.h"
```
Import the `//impeller/renderer/backend/vulkan.vk.h` umbrella header.

matanlurey replied:
Done.

chinmaygarde commented on impeller/renderer/backend/vulkan/command_pool_vk.h:
```diff
@@ -5,109 +5,126 @@
   ///             reason a valid command buffer could not be created, it will be
   ///             a `{}` default instance (i.e. while being torn down).
-  vk::UniqueCommandBuffer CreateGraphicsCommandBuffer();
+  vk::UniqueCommandBuffer CreateBuffer();
```
`CreateCommandBuffer` perhaps? There are already device buffers.

matanlurey replied:
Done.

chinmaygarde commented on impeller/renderer/backend/vulkan/command_pool_vk.h:
```diff
@@ -5,109 +5,126 @@
+///
+/// A single instance should be created per |ContextVK|.
+///
+/// Every "frame", a single |CommandPoolResourceVk| is made available for each
```
Please make a note of the the threading restrictions you've accounted for here. Specifically, the spec says:
```
Command pools are externally synchronized, meaning that a command pool must not
be used concurrently in multiple threads. That includes use via recording
commands on any command buffers allocated from the pool, as well as
operations that allocate, free, and reset command buffers or the pool itself.
```

The things I would look for are:
1. Are the commands in the command pool done executing when recycle is called?
2. Does the swap and recycle happen in a thread safe manner.

matanlurey replied:
Ack. I did find a problem (in unit tests) and addressed them. Did you want the docs updated?

matanlurey replied:
Sounds good, added!

chinmaygarde commented on impeller/renderer/backend/vulkan/command_pool_vk.h:
```diff
@@ -5,109 +5,126 @@
+  void Recycle();
+
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(CommandPoolRecyclerVK);
```
Minor Nit: These go last. The general order is. Private struct definitions -> typedefs -> private ivars -> private methods -> macros.

matanlurey replied:
Done.

chinmaygarde commented on impeller/renderer/backend/vulkan/command_pool_vk_unittests.cc:
```diff
@@ -0,0 +1,107 @@
+
+// Invokes the provided callback when the destructor is called.
+//
+// Can be moved, but not copied.
```
Can you patch `fml::ScopedAutoResetClosure` to be move safe instead?

matanlurey replied:
I could, but I want to limit how much we're changing in this PR.

Filed https://github.com/flutter/flutter/issues/134568.

chinmaygarde commented on impeller/renderer/backend/vulkan/command_pool_vk.cc:
```diff
@@ -3,166 +3,195 @@
 // found in the LICENSE file.
 
 #include "impeller/renderer/backend/vulkan/command_pool_vk.h"
+#include <memory>
```
Minor Nit: Separate by newlines; Primary header -> c sytem headers -> c++ system headers -> c engine headers -> c++ engine headers.

matanlurey replied:
Done.

chinmaygarde commented on impeller/renderer/backend/vulkan/command_pool_vk.cc:
```diff
@@ -3,166 +3,195 @@
+#include "fml/thread_local.h"
+#include "fml/trace_event.h"
+#include "impeller/renderer/backend/vulkan/resource_manager_vk.h"
+#include "vulkan/vulkan_structs.hpp"
```
Like commented elsewhere, include the `vk.h` umbrella headers. For Vulkan, VMA, OpenGL, EGL, etc..

matanlurey replied:
Done.

matanlurey commented:
I ran into some multi-threading problems with `CreateMockVulkanContext`, so I'm going to send a patch to @gaaclarke on that first before addressing the comments above (likely tomorrow AM at this point). Thanks!!!

matanlurey commented:
I found two other sources of bugs (other than code in this PR) blocking and have fixes out:

- https://github.com/flutter/engine/pull/45686
- https://github.com/flutter/engine/pull/45687

Once those are both in I'll update to main and address @chinmaygarde's comments/re-run CI.

matanlurey commented:
Ready for review!

jonahwilliams commented on impeller/renderer/backend/vulkan/command_pool_vk.cc:
```diff
@@ -4,165 +4,197 @@
+}
 
-  buffers_to_collect_.clear();
+void CommandPoolRecyclerVK::Recycle() {
```
Seems more like disposal than recycling?

matanlurey replied:
Ah fair enough. Renamed!

jonahwilliams reviewed (approved):
Overall LGTM, though not to CI 😄

matanlurey commented on impeller/renderer/backend/vulkan/command_pool_vk.cc:
```diff
@@ -4,165 +4,197 @@
-    buffer.release();
+std::optional<vk::UniqueCommandPool> CommandPoolRecyclerVK::Reuse() {
+  // If there are no recycled pools, return nullopt.
+  Lock _(recycled_mutex_);
```
Sounds good! Done.

jonahwilliams commented on impeller/renderer/backend/vulkan/fence_waiter_vk.cc:
```diff
@@ -107,6 +107,33 @@ void FenceWaiterVK::Main() {
+        break;
+      }
+
+      const auto& device = device_holder->GetDevice();
+      auto result = device.waitForFences(
+          fences.size(),                           // fences count
+          fences.data(),                           // fences
+          false,                                   // wait for all
+          std::chrono::nanoseconds{100ms}.count()  // timeout (ns)
+      );
+      if (!(result == vk::Result::eSuccess || result == vk::Result::eTimeout)) {
+        VALIDATION_LOG << "Fence waiter encountered an unexpected error "
+                       << "while terminating.";
+      }
+
```
This isn't quite right, because when we call wait for fences, not all of them will necessarily signal. Instead we need to also check if wait_set_ is empty - not the local copy of wait_set_.

jonahwilliams replied:
basically right now we have a set of N fences. We give it to waitForFences and that returns once at least one fence has updated (or timed out). But not necessarily all of the fences. So what the code is doing below is querying the fence status to determine which of the fences has updated - and remove that from the wait set. But that doesn't necessarily mean that all fences are done.

matanlurey replied:
Hmm interesting! Ok, I'll give it a look tomorrow!

matanlurey replied:
Is the change literally just `wait_set_` versus `wait_set`, or are you alluding to more?

jonahwilliams replied:
You are breaking too early. .waitForFences returning does not mean that all of the fences have signaled

matanlurey replied:
Reverted
